README.TXT
                                  Addressbook

                      © 1995, 1996, 1997 [1]Clemens Durka

         [2]Index | [3]Install | [4]Manual | [5]Screenshot | [6]Changes

     Latest version of this page at: [7]http://home.pages.de/~addressbook/

   News about Addressbook?

   Addressbook 0.7 has been released the 02.11.1997. In this version I merged
   contributions I received from users like support for more languages.

   What is Addressbook?

   Addressbook is a program to manage your addresses, phone numbers,
   e-mailaddresses and everything which is now written in your rolodex.
   Addressbook is copyrighted by [8]Clemens Durka, but available under the
   GNU GPL.

   Where can I get the program?

   For the latest release, you can always check my WWW page:
   [9]http://home.pages.de/~addressbook/

   The last public release is available at the following sites:
   [10]ftp://ftp.leo.org/pub/comp/os/linux/local/apps/addressbook-[version].tar.gz
   [11]ftp://sunsite.unc.edu/pub/Linux/X11/xapps/rolodex/addressbook-[version].tar.gz
   [12]ftp://ftp.neosoft.com/tcl/code/addressbook-[version].tar.gz

   How do I install and use Addressbook?

   • [13]INSTALL Here is the installation instruction.
   • [14]MANUAL Here is the manual.
   • [15]SCREENSHOT Here is a screenshot.

     What are the features of Addressbook?

     Some features are:

   • Graphical User Interface
   • Multi language support (now english, german, french, dutch, swedish,
     finnish)
   • All the addresses are displayed in a listbox
   • The actual address can be edited in a form
   • Configurable fields contents
   • fast search by typing the first letters of a name
   • search configurable for regular expression, matching or exact
   • External programs can be called with contents of some fields (so you can
     dial a phonenumber by pressing a button)
   • Standalone perl program for fast search without GUI

     Future features:

   • More ways and better printing
   • Export and import from databases
   • Add more languages (if some volunteer translates)

     What is needed to run Addressbook?

     Addressbook is based on Tcl/Tk, so you need a machine where Tcl/Tk is
     running. For most Unix like machines, Tcl/Tk is available. This program
     should run under any version from tcl7.3/tk3.6 to the actual version
     (tcl7.6 / tk4.2).

     This program is tested under:
   • Linux 1.2.13 (i586) Tcl7.3/Tk3.6 from Slackware 2.3
   • Linux 1.2.13 (i586) Tcl7.4/Tk4.0 from Slackware 3.0
   • Linux 2.0.29 (i586) Tcl7.6/Tk4.2 from Debian 1.3
   • HPUX 9.0x (HP 7xx)

     It is reported to work as well under:
   • HPUX 10.x (HP 7xx)
   • SunOS 4.x.x (Sun Sparc)
   • Solaris 2.4 (Sun Sparc)
   • OSF/1 3.0 (DEC Alpha)
   • AIX 3.2 (IBM Power Risc)
   • Irix 5.3 (SGI)

     Who else contributed to this package?

     Thanks to LEA (Lehrstuhl fuer Effiziente Algorithmen) of the Technical
     University of Munich and especially to Prof. Dr. E. Mayr. During a job,
     I do for them, I wrote this program.

     Thanks a lot to all the people who sended bug reports or made
     suggestions. Thanks especially to Andre Koster (scott@luna.nl) for the
     translation to dutch and to Stephen Mueller (stephen@sei.com) for
     testing and finding bugs. Thanks to Martin Paul
     (Martin.Paul@par.univie.ac.at) for contributing the mktable program for
     better printing. Thanks to all the translators for finnish (Jarmo
     Karvonen), swedish (Peter Antman) and italian (Andrea Borgia).

     Anything else?

     Questions, comments, suggestions, complaints, assistance:
     [16]clemens@dagobah.de, subject should begin with "addressbook".

   ══════════════════════════════════════════════════════════════════════════

      [17]Clemens Durka, 02.11.1997

References

   Visible links
   1. http://home.pages.de/~clemens/
   2. file:///mnt/sources/0-repos/world/museoa/addressbook/addressbook-0.7/
   3. file:///mnt/sources/0-repos/world/museoa/addressbook/addressbook-0.7/install.html
   4. file:///mnt/sources/0-repos/world/museoa/addressbook/addressbook-0.7/manual.html
   5. file:///mnt/sources/0-repos/world/museoa/addressbook/addressbook-0.7/screenshot.html
   6. file:///mnt/sources/0-repos/world/museoa/addressbook/addressbook-0.7/changes.html
   7. http://home.pages.de/~addressbook/
   8. http://home.pages.de/~clemens/
   9. http://home.pages.de/~addressbook/
  10. ftp://ftp.leo.org/pub/comp/os/linux/local/apps/addressbook-0.7.tar.gz
  11. ftp://sunsite.unc.edu/pub/Linux/X11/xapps/rolodex/addressbook-0.7.tar.gz
  12. ftp://ftp.neosoft.com/tcl/code/addressbook-0.7.tar.gz
  13. file:///mnt/sources/0-repos/world/museoa/addressbook/addressbook-0.7/install.html
  14. file:///mnt/sources/0-repos/world/museoa/addressbook/addressbook-0.7/manual.html
  15. file:///mnt/sources/0-repos/world/museoa/addressbook/addressbook-0.7/screenshot.html
  16. mailto:clemens@dagobah.de?subject=addressbook
  17. http://home.pages.de/~clemens/


CHANGES.TXT

 Nov 02 1997: public release 0.7

 Aug 22 1997: minor public release 0.6.2

 - updated country database (some phonecodes have changed)
 - finnish translation (thanks to Jarmo Karvonen <karvonen@dawn.joensuu.fi>)
 - swedish translation (thanks to Peter Antman)
 - italian translation (thanks to Andrea Borgia <borgia@cs.unibo.it>)

 May 02 1996: minor public release 0.6.1

 - mktable program contributed by Martin Paul for better
   printing (in the contrib directory)
 - updated manual
 - new searchmode caseinsensitive regexp
 - bugfix selection in listbox
 - fonts and color configurable
 - start as icon with -iconic
 - iconbitmap added
 - texconvert: _ to \_ and ^ to \^
 - option for not showing the ABC and action buttons

 Mar 22 1996: public release 0.6

 - bugfix in global preferences

 Mar 17 1996: test release 0.5.8

 - documentation now in HTML and in ASCII
 - bugfixes in address.perl
 - bugfixes with file locking

 Mar 12 1996: test release 0.5.7

 - file locking: only the first person gets write access
   (related options: lockbydefault, nolocking)
 - added debuggingoutput (call with -debug 'level'; 1 <= level <= 4)
 - printing with canvas works a little bit
 - state field for US users (see sample/addresses_usa.dat)
 - added a second command line program written in awk. This is not
   installed by default, but are welcome to try it.

 Feb 02 1996: test release 0.5.6

 Changes from 0.5.5 to 0.5.6

 - export works
 - included dutch translation (thanks to Andre Koster <scott@luna.nl>)
 - beginning of printing with a canvas (doesn't work yet)

 Jan 24 1996: internal release 0.5.5

 Changes from 0.5.4 to 0.5.5

 - little better printing (use of mpage instead of a2ps for postscript)
 - latex user defined printing
   (to use it, define the variables userhead and usertail with the latex
   code which comes before and after the addresses in your configfile)
 - export addresses
 - new countries added in countrydatabase
 - saving of global preferences

 Dec 11 1995: test release 0.5.4

 Changes from 0.5.3 to 0.5.4

 - bugfixes

 Nov 29 1995: internal release 0.5.3

 Changes from 0.5 to 0.5.3:

 - selection after a search remains there. You can select and unselect
   entries by doubleclick. Will be used for printing.
 - confirmwindow on each operation is configurable
   (configoption: confirm, confirmdelete)
 - undo button to undo the last add delete or update operation
 - sorting is independant from the listboxentries
   (in the .fmt file: listboxentry1 / 2  and sort1 / 2)
   if the sort is omitted, the listboxentry becomes sortingcriteria
 - added www button, if you use www field
 - fast loading of up to 3 addressfiles. You have to define in your
   configfile adrfile1 / 2 / 3 for the filename of the database and
   adrfile1alias (and 2, 3) for a name that appears in the menu.
 - many new predefined fieldnames (see samples/addresses_overload.dat)
 - every predefined fieldname can be doubled (street -> street2)
 - support for more languages (probably dutch, spanish
   if the translators help me)
 - reordered countrydatabasefile to support more languages and added
   a field for the ISO countrycode for future use.
 - bugfix in fileselectionbox for loading
 - completely bilingual for use with tk4.0 or tk3.6
 - bindings conform to tk4.0
 - cut & paste works for tk3.6 again
 - contrib directory and a simple birthday reminder in it.

 May 28 1995: public release of addressbook-0.5

 Changes from 0.4 to 0.5:

 - many bugfixes
 - some support for printing lists
 - support for tcl7.4/tk4.0
 - deleting addresses is possible
 - little test and installationscript
 - user defined fields
 - much more, I don't remember any more :-)

 Apr 03 1995: public release of addressbook-0.4 (first public release)


lib/address.part2.awk
# tel.awk 2. Teil

# Aufsplitten des Suchargumentes
split(SEARCH, S, " ")
SEARCH1=S[1]
SEARCH2=S[2]
}


# Funktion zur Texkonvertierung
function converttotex (line) {
  result=""
  i=1
  end=length(line)

  while(i <= end) {
    zeichen=substr(line,i,1)
    
    if (zeichen == "#") zeichen = "\\#"
    if (zeichen == "&") zeichen = "\\&"
    if ((zeichen == " ") && (substr(line,i-1,1) == ".")) zeichen = "~"

    result=sprintf("%s%s",result,zeichen)
#    result=zeichen
    i++
  }
  return result
}


# Funktion zur Ausgabe
function output (line) {
  if (ACTION ~ /lettertex/) {
    line=converttotex(line)
  }
  printf "%s\n", line
}

# Funktion zur Konvertierung von Telefonnummern
function prepare_number (line, country) {
    # Nummer extrahieren
    gsub(/[A-Za-z]/,"",line)
    gsub(/[ ()/\-]/,"",line)
#    match(line,"[(]*[0-9)/\-]*")
#    nummer=substr(line,RINDEX,RLENGTH)
    if (country != mycountry) {
      # international call
      tel = dialoutdistance
      tel = tel intl_dialout
      if (index(line,"+" intl_prefix[country]) == 0) {
        # kein +, also landesvorwahl anfuegen
	tel = tel intl_prefix[country]
	if (index(line,intl_leaveout[country]) == 1) { 
          # '0' oder so entfernen
	  line = substr(line,1+length(intl_leaveout[country]))
	}
      }
      tel = tel line
      gsub(/+/,"",tel)
    } else {
      # nationaler call
      if (index(line,"+" intl_prefix[country]) == 1) {
        # + Landesvorwahl entfernen
	line = intl_leaveout[country] substr(line,2+length(intl_prefix[country]))
      }
      if (index(line,myareacode) == 1) {
        # Ortsgespraech Vorwahl entfernen
	tel = dialoutlocal substr(line,1+length(myareacode))
      } else {
        # Ferngespraech
	tel = dialoutdistance line
      }
    }
    return tel
}

# Funktion zum Ausdruck formatieren 
function myprint (line, field) {
    split(line, l, ";")
    if (field == mail) {
# Adresse
	if (l[title] != "") {
	    output(l[title] " " l[firstname] " " l[lastname])
	} else {
 	    output(l[firstname] " " l[lastname])
	}
	if (l[department] != "") output(l[department])
    	if (l[institute] != "") output(l[institute])
	if (l[addon] != "") output(l[addon])
    	if (l[pobox] != "") output(l[pobox])
    	if (l[street] != "") output(l[street])
	if (l[country] == "D") { 
	    output(l[zip] " " l[city])
	} else {
	    if ((l[country] == "USA") ||(l[country] == "MEX")) {
	    	output(l[city] ", " l[zip])
	    } else {
		if ((l[country] == "GB") || (l[country] == "AUS") ||
	    	    (l[country] == "J") || (l[country] == "CDN") ||
		    (l[country] == "SGP") || (l[country] == "IND")) {
		    output(l[city] " " l[zip])
		} else {
	    	    output(l[country] "-" l[zip] " " l[city])
		}
	    }
	}
	if (l[country] != "D") {
	    output(fc[l[country]])
	}
	output("")
    } else {
# Keine Adresse
        anzahl=split(l[field],a,",")
        if (anzahl == 0) {
	  printf "Keine Information hierzu fuer %s %s vorhanden.\n", l[firstname], l[lastname] > "/dev/stderr"
	} else if (anzahl == 1) {
	  if ((field == tel) || (field == fax)) {
	    print(prepare_number(a[1],l[country]))
	  } else {
	    print a[1]
	  }
	} else {
	  for (i=1; i<=anzahl; i++) {
# Fuehrendes Leerzeichen entfernen
	    gsub(/^ /,"",a[i])
	    gsub(/\(H\)/,"(home)",a[i])
	    gsub(/\(O\)/,"(office)",a[i])
	    gsub(/\(W\)/,"(work)",a[i])
	    gsub(/\(S\)/,"(secretary)",a[i])
	    gsub(/\(P\)/,"(pager)",a[i])
	    gsub(/\(M\)/,"(mobile)",a[i])
	    gsub(/\(C\)/,"(cellular)",a[i])
	    printf "%-35s #%4s\n", a[i] ,i > "/dev/stderr"
	  }
	  printf "Mehrere Nummern fuer %s %s vorhanden. Bitte waehlen: ", l[firstname], l[lastname] > "/dev/stderr"
	  getline ans < "/dev/stdin"
	  if ((ans != " ") && (ans != "q") && (ans != "")) {
	    if ((field == tel) || (field == fax)) {
	      print(prepare_number(a[ans],l[country]))
	    } else {
	      print a[ans]
	    }
	  }
 	}
    }
}


# Hauptschleife zum Suchen
{
if (((match($(firstname), SEARCH1)) && (match($(lastname), SEARCH2))) ||
    ((match($(firstname), SEARCH2)) && (match($(lastname), SEARCH1)))) {
	i++
	found[i] = $0
    }

}
    	

# Endbehandlung Ausgabe
END {
# print SEARCH " : " i " Vorkommen "  NUM

if (i == 0) {
    print SEARCH " : nicht gefunden"
    exit 1
}

if (ACTION ~ /tel/) { field=tel }
if (ACTION ~ /fax/) { field=fax }
if (ACTION ~ /email/) { field=email }
if (ACTION ~ /letter/) { field=mail }

if (i == 1) {
# Nur ein Vorkommen
    myprint(found[1],field)
} else {
    if (NUM != "") {
# Mehrere Vorkommen, eins wurde ausgewaehlt
	myprint(found[NUM],field)
    } else {
# Mehrere Vorkommen, alle werden mit Namen angezeigt.
        # print "Mehrere Vorkommen gefunden." > "/dev/stderr"
	for (j=1; j<=i; j++ in found) {
	    split(found[j], f, ";") 
	    printf "%-35s #%4d  %s %s, %s\n", f[field], j,
	    f[firstname],f[lastname],f[city] > "/dev/stderr"
	}
	printf "Bitte passende Nummer eingeben (q oder SPACE = quit): " > "/dev/stderr"
	getline ans < "/dev/stdin"
	if ((ans != " ") && (ans != "q") && (ans != "")) {
	  myprint(found[ans],field)
	}
    }
}

}


lib/create_countryinfo.awk
#!/usr/bin/gawk -f
#
#########################################################################
#                                                                       #
# Creates the file countryinfo.awk with information from countries	#
# Call gawk -f makecountryinfo.awk countries > countryinfo.awk		#
#                                                                       #
# This is part of my adressbuch / addressbook program                   #
# Version 0.7, 02.11.1997                                               #
# Copyright (C) 1995, 1996, 1997 Clemens Durka                          #
#                                                                       #
# Clemens Durka (clemens@dagobah.de)                                    #
# http://home.pages.de/~clemens/                                        #
#                                                                       #
#########################################################################
#
BEGIN {FS=";"; OFS=" "
print "# Countryinformation"
print "#"
print "### DO NOT EDIT ###"
print "#"
print "# This file has been created automatically from countries"
print "# If you need to change anything, edit countries and contact the author."
print "#"

}

/^#/ { next }

{
print "fc[\"" $1 "\"]=\"" $9 "\""
print "intl_prefix[\"" $1 "\"]=\"" $3 "\""
print "intl_leaveout[\"" $1 "\"]=\"" $5 "\""
}


contrib/mktable-1.0/parse.c
/*
** $Id: parse.c,v 1.1 1996/04/17 14:03:14 martin Exp $
** 
** $Log: parse.c,v $
** Revision 1.1  1996/04/17 14:03:14  martin
** Initial revision
**
*/

#include <stdio.h>
#include <string.h>
#include <ctype.h>

#include "parse.h"

/* Print fatal error and exit */
#define FATAL(MSG) { fprintf (stderr, "%s\n", MSG); exit (1); }

/* Print parse error and exit */
#define PARSE_ERROR(MSG) {fprintf (stderr, "Parser error: %s (pos %d)\n", MSG, fi); exit(1);}

/* Justification */
#define LEFT 1
#define RIGHT 2

#define MAXNUMS 25

/* To make the parser more readable */
#define INCH format[fi]
#define RESET fi=0
#define FWD fi++


static char *format;
static int nums[MAXNUMS];
static int nums_nr;
static int length;
static int just;
static int fi;

/* Local function prototypes */

static void parse_text(void);
static void parse_field(void);
static int parse_num();


/*
 * Function definitions
 */

/* 
 * Prints one entry with the format in format_p
 */
extern void
print_entry(char format_p[], char *entry[], int field_c)
{
	int i;
	char con[1024];
	char fmt[1024];

	format= format_p;
	RESET;
	
	for (;;) {
		if (INCH == 0)
			break;
		if (INCH == '@') {
			parse_field();

			*con=0;
			for (i=0; i< nums_nr; i++) {
				if (nums[i] >= field_c) {
					fprintf (stderr, 
						"Trying to access field %d (of %d)\n",
						nums[i], field_c
					);
					exit (1);
				}
				if (strlen(con) != 0)
					strcat (con, " ");
				strcat (con, entry[nums[i]]);
			}
			if (length != -1) {
				if (strlen(con) > length) {
					con[length-1]= '!';
					con[length]= 0;
				}
			}
			/* Build the format string for current printfield */
			fmt[0]= 0;
			strcat (fmt, "%");
			if (length != -1)
				sprintf (fmt, "%s%d", fmt, length);
			if (just == LEFT)
				strcat (fmt, "-");
			strcat (fmt, "s");

			printf (fmt, con);
		} else {
			parse_text();
		}
	}
}


/* Parse text from the input and put it to stdout */
static void
parse_text(void)
{
	while (INCH != '@' && INCH != 0) {
		putchar (INCH);
		FWD;
	}
}


/* Parses a field definition (@1[...) */
static void
parse_field(void)
{
	FWD;

	length=-1;
	just=LEFT;

	if (!isdigit(INCH))
		PARSE_ERROR ("expected field number");

	nums_nr=0;
	nums[nums_nr++]= parse_num();
	while (INCH=='+') {
		FWD;
		nums[nums_nr++]= parse_num();
	}
	if (INCH != '[')
		return;

	FWD;
	if (isdigit(INCH)) {
		length= parse_num();
	}
	if (INCH=='L') {
		just=LEFT;
		FWD;
	} else if (INCH=='R') {
		just=RIGHT;
		FWD;
	} 
	if (INCH==']') {
		FWD;
		return;
	}
	if (INCH == 0) {
		PARSE_ERROR("unexpected EOF");
		return;
	}
	PARSE_ERROR("unexpected character after length param");
}


/* Parses a number (1 or more digits) */
static int
parse_num()
{
	int num;

	num=INCH-'0';
	FWD;
	while (isdigit (INCH)) {
		num *= 10;
		num += INCH-'0';
		FWD;
	}
	if (INCH==0)
		PARSE_ERROR ("unexpected EOF");
	return (num);
}


contrib/mktable-1.0/mktable.c
/*
** $Id: mktable.c,v 1.2 1996/04/17 14:07:39 martin Exp $
** 
** $Log: mktable.c,v $
** Revision 1.2  1996/04/17 14:07:39  martin
** Now searching in other directories for the format files as well
** Using the new parser in parse.c
**
** Revision 1.1  1996/04/17 09:24:06  martin
** Initial revision
**
*/

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <ctype.h>
#include <sys/param.h>

#include "parse.h"

#define TRUE 1
#define FALSE 0

/* Print a fatal error message and exit */
#define FATAL(MSG) { fprintf (stderr, "%s\n", MSG); exit (1); }

/* formats for the input files */
#define FILEFMT_HEAD "%s/%s.head"
#define FILEFMT_ENTRY "%s/%s.entry"
#define FILEFMT_FOOT "%s/%s.foot"

/* The default directory to search for formats */
#define DEFAULT_FORMAT_PATH "/opt/local/lib/mktable"

/* This env. var. can hold a dir where formats should be searched for */
#define FORMAT_PATH_ENVVAR "MKTABLEFORMATS"

/* Maximum number of fields and lines */
#define MAXFIELDS 25
#define MAXENTRIES 4096

/* The data from the input file */
char *entries[MAXENTRIES][MAXFIELDS];

/* The number of entries in the database (tuples, lines) */
static int entry_c;

/* Fields per tuple */
static int field_c;

/* The delimiter used to seperate the fields. Can be overridden on cmdline */
static int delim=';';

/* The format string is read into this var */
char format[4096];

/* The name pf the format template */
static char formatf[MAXPATHLEN];

/* The name of the input file (if NULL: stdin is read) */
static char inputf[MAXPATHLEN];

/* if TRUE: Only check the input file (don't generate a table) */
static int check_only=FALSE;


/* Local prototypes */

static void eval_arguments(int argc, char *argv[]);
static void read_table(char *fname);
static int read_file(char *fname, char str[]);
static void print_file(char *fname);
static void print_stat(void);
static void usage(void);
static char *new_strtok(char *s1, char sep);
static char * find_format_path(char *formatf);


/*
 * Function definitions
 */
	
extern int
main(int argc, char *argv[])
{
	int i;
	char fname[MAXPATHLEN];
	char *format_path;

	/* Read command line args */
	eval_arguments(argc, argv);

	/* Read the data file into memory */
	read_table(inputf);

	/* Check only ? print statistics and exit */
	if (check_only==TRUE) {
		print_stat();
		exit (0);
	}

	format_path= find_format_path(formatf);

	/* Read the format specs in the .entry file */
	sprintf (fname, FILEFMT_ENTRY, format_path, formatf);
	read_file(fname, format);

	/* Print header */
	sprintf (fname, FILEFMT_HEAD, format_path, formatf);
	print_file(fname);

	/* Print entries */
	for (i=0; i<entry_c; i++) {
		print_entry(format, entries[i], field_c);
	}

	/* Print foot lines */
	sprintf (fname, FILEFMT_FOOT, format_path, formatf);
	print_file(fname);

	exit (0);
}


/*
 * Evaluate command line arguments
 */
static void
eval_arguments(int argc, char *argv[])
{
	extern char *optarg;
	extern int optind;

	int c;

	while ((c=getopt(argc, argv, "d:f:c")) != -1) {
		switch (c) {
		case 'd':
			delim= optarg[0];
			break;
		case 'f':
			strcpy (formatf, optarg);
			break;
		case 'c':
			check_only=TRUE;
			break;
		case '?':
			usage();
		}
	}
	/* Do we have a file name for the data file ? */
	if (optind == argc-1) {
		strcpy (inputf, argv[optind]);
		return;
	}
	/* If there is no filename, we'll read stdin */
	if (optind == argc) {
		*inputf= 0;
		return;
	}

	/* There are arguments left which we cannot parse -> error */
	usage();
}


/*
 * Print usage information 
 */
static void
usage(void)
{
	fprintf (stderr, "usage:    mktable [-d<delim>] -f<format> [datafile]\n");
	fprintf (stderr, "       or mktable [-d<delim>] -c [datafile]\n");
	fprintf (stderr, "\n");
	fprintf (stderr, "Options are:\n");
	fprintf (stderr, "  -d<delim>   field delimiter (one char)\n");
	fprintf (stderr, "  -f<format>  The format name.\n");
	fprintf (stderr, "  -c          Print some stats on the input file and exit.\n");
	fprintf (stderr, "  [datafile]  The table file. If not given, stdin is read.\n");

	exit (1);
}


/* 
 * Read the data file into memory 
 */
static void
read_table(char *fname)
{
	FILE *in;
	char line[1024+1];
	char *start;
	int cur_field_c;
	int i;

	/* If we have no input file name, we use stdin */
	if (*inputf==0) {
		in= stdin;
	} else {
		in= fopen(fname, "r");
		if (in == NULL)
			FATAL ("Could not open input file");
	}
	
	/* Count entries (tuples) while reading them */
	entry_c=0;
	for (;;) {
		/* EOF ? */
		if (fgets (line, 1024, in) == NULL)
			break;

		/* Ignore empty lines */
		if (strlen (line) == 0)
			continue;

		/* Cut trailing newline */
		line[strlen(line)-1]= 0;

		/* Read all the fields in one line, count fields in cur_field_c */
		cur_field_c= 0;
		start= new_strtok (line, delim);
		while (start != NULL) {
			entries[entry_c][cur_field_c]= strdup(start);
			cur_field_c++;
			start= new_strtok (NULL, delim);
		}

		/* The number of fields in the first entry will be the
		   standard fieldcount.
		*/
		if (entry_c == 0)
			field_c= cur_field_c;

		/* If the current line has less fields than the standard line,
		   fill the missing fields with empty strings.
		   If the current line has MORE fields than the standard line,
		   these fields are ignored. Otherwise we would have to adapt
		   all other entries .. normally all this should not happen,
		   because every line should have the same number of fields ..
		*/
		if (cur_field_c < field_c) {
            for (i=cur_field_c; i<field_c; i++)
                entries[entry_c][i]= strdup ("");
		}
		/* And on to the next entry (line) .. */
		entry_c++;

		/* Check for overflow in the input file */
		if (entry_c > MAXENTRIES) {
			fprintf (stderr, "Max number of entries reached (%d) -- Exit.\n", MAXENTRIES);
			exit (1);
		}
	}
}


/*
 * Prints the contents of a file to stdout.
 * If the file is not existant or it is empty, nothing is printed,
 * this is not an error.
 */
static void
print_file(char *fname)
{
	FILE *in;
	int ch;

	in=fopen(fname, "r");
	if (in == NULL)
		return;

	while ((ch=fgetc (in)) != EOF)
		putchar (ch);

	fclose (in);
}


/*
 * Reads the format specs from the .entry file
 * More generally, the file <fname> is read into the string <str>
 */
static int
read_file(char *fname, char str[])
{
	FILE *in;
	int ch;
	int i;

	/* Open file, if not possible - fatal error */
	if ((in=fopen(fname, "r")) == NULL) {
		fprintf (stderr, "Could not open file <%s>\n", fname);
		exit (1);
	}

	i=0;
	while ((ch=fgetc (in)) != EOF) {
		str[i]= ch;
		i++;
	}
	str[i]=0;

	fclose (in);
	return (i);
}


/*
 * A little different from strtok(). All fields are seperated by
 * exactly one sep. char, and a seperator at the end means an
 * empty field following.
 */
static char *
new_strtok(char *s1, char sep)
{
	static char *el;
	static int last;
	char *hlp;
	char *ret;

	if (s1 != NULL) {
		el= s1;
		last= FALSE;
	}
	if (last == TRUE)
		return (NULL);

	hlp= el;
	while (*hlp != sep && *hlp != 0)
		hlp++;

	if (*hlp == 0)
		last=TRUE;

	*hlp= 0;
	ret= el;
	el= ++hlp;
	return (ret);
}


/*
 * Prints some statistics about the input data file.
 */
static void
print_stat(void)
{
	int i, j;
	int max_len, max_idx;

	/* Show number of entries and fields/entry */
	printf ("Got %d entries, %d fields each\n", entry_c, field_c);

	/* Show, for each field, the length of the longest value */
	printf ("Maximum lengths of field values:\n");
	for (i=0; i<field_c; i++) {
		max_len=0;
		max_idx=0;
		for (j=0; j<entry_c; j++) {
			if (strlen (entries[j][i]) > max_len) {
				max_len= strlen (entries[j][i]);
				max_idx= j;
			}
		}
		printf ("Field %2d: %3d (%s)\n", i, max_len, entries[max_idx][i]);
	}
}


/*
 * Find the path where .entry, .head, ... are to be searched for
 */
static char *
find_format_path(char *formatf)
{
	char fullname[MAXPATHLEN];

	/* The current working dir ? */
	sprintf (fullname, FILEFMT_ENTRY, ".", formatf);
	if (fopen(fullname, "r") != NULL)
		return (".");

	/* .. or in the default format path (can be defined at compile time) */
	sprintf (fullname, FILEFMT_ENTRY, DEFAULT_FORMAT_PATH, formatf);
	if (fopen(fullname, "r") != NULL)
		return (DEFAULT_FORMAT_PATH);

	/* .. or in the dir from an environment variable whose name
	   is stored in FORMAT_PATH_ENVVAR
	 */
	if (getenv(FORMAT_PATH_ENVVAR) != NULL) {
		sprintf (fullname, FILEFMT_ENTRY, getenv(FORMAT_PATH_ENVVAR), formatf);
		if (fopen(fullname, "r") != NULL)
			return (getenv (FORMAT_PATH_ENVVAR));
	}
	
	/* The format was found nowhere .... */
	fprintf (stderr, "Could not find format <%s>\n", formatf);
	exit (1);
}
